packages/rpc-core/src/rpc.ts

```diff
@@ -0,0 +1,22 @@
+import { IJsonRpcTransport } from '@solana/rpc-transport';
+import { JsonRpcApi } from './types/jsonRpcApi';
+
+export const rpc = /* #__PURE__ */ new Proxy<JsonRpcApi>({} as JsonRpcApi, {
+    defineProperty() {
+        return false;
+    },
+    deleteProperty() {
+        return false;
+    },
+    get<TMethodName extends keyof JsonRpcApi>(target: JsonRpcApi, p: TMethodName) {
+        if (target[p] == null) {
+            const method = p.toString();
+            target[p] = async function (transport: IJsonRpcTransport, ...params: Parameters<JsonRpcApi[TMethodName]>) {
+                const normalizedParams = params.length ? params : undefined;
+                const result = await transport.send(method, normalizedParams);
+                return result;
+            } as unknown as JsonRpcApi[TMethodName];
```

Comment:
Learning a lot today with `... as unknown as ...` -- does this cause potential footguns for transport implementers?

Reply:
At that point, it's our job to make sure that the implementation of `rpc.*` is perfect. Essentially what we're saying is that the dynamic `get` returns a method that will definitely conform to `JsonRpcApi[TMethodName]` but the implementation on lines 14-18 does not, from Typescript's perspective.

```shell
src/rpc.ts:14:13 - error TS2322: Type '(transport: IJsonRpcTransport, ...params: Parameters<JsonRpcApi[TMethodName]>) => Promise<unknown>' is not assignable to type 'JsonRpcApi[TMethodName]'.
  Type '(transport: IJsonRpcTransport, ...params: Parameters<JsonRpcApi[TMethodName]>) => Promise<unknown>' is not assignable to type '{ (transport: IJsonRPCTransport, address: Base58EncodedAddress, config?: ({ encoding: "base64"; } & GetAccountInfoApiCommonConfig & GetAccountInfoApiBase64EncodingCommonConfig) | undefined): Promise<...>; (transport: IJsonRPCTransport, address: Base58EncodedAddress, config?: ({ ...; } & ... 1 more ... & GetAccountIn...'.
    Type '(transport: IJsonRpcTransport, ...params: Parameters<JsonRpcApi[TMethodName]>) => Promise<unknown>' is not assignable to type '{ (transport: IJsonRPCTransport, address: Base58EncodedAddress, config?: ({ encoding: "base64"; } & GetAccountInfoApiCommonConfig & GetAccountInfoApiBase64EncodingCommonConfig) | undefined): Promise<...>; (transport: IJsonRPCTransport, address: Base58EncodedAddress, config?: ({ ...; } & ... 1 more ... & GetAccountIn...'.
      Types of parameters 'params' and 'address' are incompatible.
        Type '[address: Base58EncodedAddress, config?: ({ encoding: "base64"; } & GetAccountInfoApiCommonConfig & GetAccountInfoApiBase64EncodingCommonConfig) | undefined]' is not assignable to type 'Parameters<JsonRpcApi[TMethodName]>'.

14             target[p] = async function (transport: IJsonRpcTransport, ...params: Parameters<JsonRpcApi[TMethodName]>) {
               ~~~~~~~~~
```

Reply:
It _seems_ like these two should be compatible:

* `[address: Base58EncodedAddress, config?: ({ encoding: "base64"; } & GetAccountInfoApiCommonConfig & GetAccountInfoApiBase64EncodingCommonConfig) | undefined]`
* `Parameters<JsonRpcApi[TMethodName]>`

…but they aren't.

Reply:
Ah I see, that's really weird, thanks for explaining. Is that a typescript bug? Or does `Parameters` need to have some extra declaration to make that clear?

Reply:
I'm not sure, but I think I made a mistake including `transport` in all the method signatures. I wonder what would happen if I took that out…

Reply:
#1196.